feature/dynamic-controls

tzinckgraf opened the pull request:
- added dynamic controls based on the config file
- replaced the controls in the story
- only replaced the FlyTo control so far

I have a config set up similar to the following
```
  "layers": [
    {
     "property": {
        "id": "StatueOfLiberty",
        "name": "Statue of Liberty",
        "type": "geojson",
        "url": "layers/statue-of-liberty.geojson"
      },
      "controlProperties": [
        {
          "id": "flyTo",   
          "type": "flyTo", 
          "subtitle": "This is the Statue of Liberty",
          "args": {                                                                                      
            "lat": 40.68967735303955,
            "lng": -74.04534588323135,
            "zoom": 15
          }
        },
        {  
          "id": "info",
          "subtitle": "This is some information on the Statue of Liberty"
        }                                                                                      
      ]                                                                                        
    },
....    
```

This creates two controls. One is a fly to button. The other is just a read only component with the subtitle.

tzinckgraf committed:
feature/dynamic-controls
- added dynamic controls based on the config file
- replaced the controls in the story
- only replaced the FlyTo control so far

tzinckgraf commented on src/components/Story.svelte:
```diff
@@ -1,42 +1,28 @@
-</div>
 
 <!-- hacky solution until I learn about the svelte css better. Otherwise, the if ignores the class -->
 <div class="panel">
```
I am removing the slider next, but one control at a time.

tzinckgraf committed:
feature/dynamic-controls
- fixed config to use generic arguments

njason commented on src/components/Story.svelte:
```diff
@@ -1,42 +1,28 @@
+  onMount(async () => {
+      const controlPromises = config.layers
+        .map((layerConfig: Layer) => layerConfig.controlProperties?.map(l => {
+            return {idLayer: layerConfig.id, nameLayer: layerConfig.name, controlProperties: l}}))
```
Why not just continue using the `Layer` type instead of making a new one?

tzinckgraf replied:
As discussed, each layer can have multiple controls. I could have just passed the whole layer config to each control, but I didn't want a control to know the configs to all other controls.

As I build out these layers, I am thinking of updating that Layers config to look like
```
export interface Layer {    
  id: string,
  property: LayerProperty,
  controlProperties?: ControlProperty[]
  args?: [a better typed args]    
}  
```
This means the Layer config is really just a mapping, and it would clean up this code to let me pass those properties along to each control. The controls will need those properties to do things like find the correct stores.

Thoughts?

njason replied:
What about `const controlPromises = config.layers.map((layerConfig: Layer) => CreateControls(layerConfig)).flat()`

What's in `LayerProperty`?

tzinckgraf replied:
That would work. I will make the change. Thank you for the suggestion.

`LayerProperty` would for now just include the name and the url, but it would give us the flexibility of adding more properties in the future. The problem now, and it is creating some ugly code, is that a control may need to know details about the layer. The name is a great example here. If we pushed that into a `LayerProperty`, then it cleans up for future versions. I can push a fix for that in this PR to clean up this code.

njason commented on src/config.ts:
```diff
@@ -16,12 +16,16 @@ export interface Layer {
-  prefixKey?: string
+  prefixKey?: string,
+  // generic arguments that depends on the type
+  args: any
```
I would really like to avoid using `any`

tzinckgraf replied:
I have a fix in another branch to remove this. We need to add some more pieces to this config, so I started splitting it out to arg types by control type (for example, a slider has slider specific arguments). In an effort to keep this PR small, can we keep this `any` and I will remove it in one of my next PRs?

njason reviewed (commented):
I don't think flyTo should be a control property since it doesn't control the layer. The layer's control properties should deal with changing the presentation of the layer, like the income slider you had.

I can make my proposal for flyTo in the next PR I make which was going to add the stories to the UI from config.

tzinckgraf committed:
feature/dynamic-controls
- split out layer properties to a separate object
- moved logic to control creator

tzinckgraf commented on src/components/controls/ControlCreator.svelte:
```diff
@@ -0,0 +1,35 @@
+    //import Slider from './Slider.svelte';
+    import type { Layer, ControlProperty, LayerProperty } from '../config';
+
+    // NOTE: this is async in antication for dynamically loading components. that feature is async.
```
FIXME: typo here.

tzinckgraf commented:
I am going to close this to redo the PR and address all the above requests.
